wallet/wallet.go

```diff
@@ -74,120 +66,76 @@ type walletSecretsEncrypted struct {
-func NewWallet() (*Wallet, error) {
-	return NewMultiWallet(1)
+	return NewMultiWalletFromMnemonic(m, n)
 }
 
-// NewMultiWallet creates a new wallet containing multiple accounts generated using random seeds.
-func NewMultiWallet(n int) (*Wallet, error) {
+func NewMultiWalletFromMnemonic(m string, n int) (*Wallet, error) {
 	if n < 0 || n > common.MaxAccountsPerWallet {
 		return nil, fmt.Errorf("invalid number of accounts")
 	}
-	var accounts []*EDKeyPair
-	for i := 0; i < n; i++ {
-		a, err := NewRandomAccount()
-		if err != nil {
-			return nil, err
-		}
-		accounts = append(accounts, a)
+	if !bip39.IsMnemonicValid(m) {
+		return nil, fmt.Errorf("invalid mnemonic")
 	}
-	return walletFromAccounts(accounts)
-}
-
-func NewRandomAccount() (*EDKeyPair, error) {
-	e, err := bip39.NewEntropy(ed25519.SeedSize * 8)
+	// TODO: add option for user to provide passphrase
+	// https://github.com/spacemeshos/smcli/issues/18
+	seed := bip39.NewSeed(m, "")
+	masterKeyPair, err := NewMasterKeyPair(seed)
 	if err != nil {
 		return nil, err
 	}
-	return accountFromSeed(e)
-}
-
-// NewWalletFromSeed creates a new wallet containing one account generated using the given seed.
-func NewWalletFromSeed(seed []byte) (*Wallet, error) {
-	kp, err := accountFromSeed(seed)
+	accounts, err := accountsFromMaster(masterKeyPair, seed, n)
 	if err != nil {
 		return nil, err
 	}
-	return walletFromAccounts([]*EDKeyPair{kp})
+	return walletFromMnemonicAndAccounts(m, masterKeyPair, accounts)
 }
 
-func walletFromAccounts(kp []*EDKeyPair) (*Wallet, error) {
-	displayName := "Main Wallet"
-	createTime := common.NowTimeString()
-
+func walletFromMnemonicAndAccounts(m string, masterKp *EDKeyPair, kp []*EDKeyPair) (*Wallet, error) {
 	w := &Wallet{
 		Meta: walletMetadata{
-			DisplayName: displayName,
-			Created:     createTime,
+			DisplayName: "Main Wallet",
+			Created:     common.NowTimeString(),
 			// TODO: set correctly
 			GenesisID: "",
-			Meta: MetaMetadata{
-				Salt: common.DefaultEncryptionSalt,
-			},
 		},
 		Secrets: walletSecrets{
-			//Mnemonic: mnemonic,
-			Accounts: kp,
+			Mnemonic:      m,
+			MasterKeypair: masterKp,
+			Accounts:      kp,
 		},
-		//masterKeypair: masterKeyPair,
 	}
 	return w, nil
 }
 
-// accountFromSeed creates a new account from the given seed.
-func accountFromSeed(seed []byte) (*EDKeyPair, error) {
-	masterKeyPair, err := NewMasterKeyPair(seed[:ed25519.SeedSize])
-	if err != nil {
-		return nil, err
+// accountsFromMaster generates one or more accounts from a master keypair and seed. Accounts use sequential HD paths.
+// The master keypair does not contain the seed that was used to generate it, so it needs to be passed in explicitly.
+func accountsFromMaster(masterKeypair *EDKeyPair, masterSeed []byte, n int) (accounts []*EDKeyPair, err error) {
+	for i := 0; i < n; i++ {
+		acct, err := masterKeypair.NewChildKeyPair(masterSeed, i)
+		if err != nil {
+			return nil, err
+		}
+		accounts = append(accounts, acct)
```

Comment:
`append` is costly, since the slice needs to be copied on every call if there is no additional capacity. Instead you can pre-allocate a slice with the right capacity to avoid copying data:

```suggestion
	accounts = make([]*EDKeyPair, 0, n)
	for i := 0; i < n; i++ {
		acct, err := masterKeypair.NewChildKeyPair(masterSeed, i)
		if err != nil {
			return nil, err
		}
		accounts = append(accounts, acct)
```